### PR TITLE
Dashboard: Add API hooks for categories, tags, users. Connected the data fetched to the story's list UI.

### LINKED ISSUE #770: Dashboard - List View UI
## Feature Description

As an author I want to see stories in an elegant way

**[Feature Brief](https://docs.google.com/document/d/1Xs4uOZ3KCNlTMdjDT6q7cvTUpA1xk-kzrnVLQALltQk/edit#heading=h.mgdx02j3ibol)**
**[Design](https://www.figma.com/file/hUTw97921an0e3uU28JoDn/Dashboard?node-id=1853%3A93070)** Note as of Mar 25, Sam will be adding a "simple" view to figma

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance Criteria
1. Build the UI for the list view

### PR DESCRIPTION
**Pull request by carloskelly13**
## Summary

Adds the ability to fetch the users, categories, and tags from the Wordpress API. Matches up the ids from the story to the user-facing strings from the data.

Screen Shot 2020-04-30 at 5 12 01 PM

## Relevant Technical Choices

- Follow the existing `useXApi` pattern to create smaller, testable modules for fetching and storing data
- Creating minimum width columns for author and last modified to ensure the header title doesn't crop

## To-do

- [x] Add in the two new columns from the new design videos in the next PR
Fixes #770

### CODE REVIEW

File: assets/src/dashboard/app/views/shared/storyListView.js [hunk 1]
```diff
@@ -92,13 +97,33 @@ const toggleSortLookup = {
...
+  tags,
+  categories,
+  users,
 }) {
+  const metadataStringForIds = useCallback((metadata, ids) => {
+    const metadataString = ids
+      .reduce((memo, current) => {
+        return [...memo, metadata[current].name];
+      }, [])
+      .join(', ');
```
Comment (BrittanyIRL) [inline assets/src/dashboard/app/views/shared/storyListView.js]: To test this, I went through my local wordpress set up and added a handful of tags and categories. This reduce is getting caught by having an id that doesn't exist in the metadata and searching for its name. A story of mine has 3 tags, but the tags returned through the api is missing one of those Ids and causing the callback to error out on the reduce.  This is my data, if it helps - but I think just doing a check that the name exists would be good. 

From stories object: 
```
{
22:
author: 1
bottomTargetAction: "http://localhost:8899/wp-admin/post.php?action=edit&post=22"
categories: (2) [9, 11]
centerTargetAction: ""
id: 22
modified: Moment {_isAMomentObject: true, _i: "2020-05-01T15:48:27", _f: "YYYY-MM-DDTHH:mm:ss", _isUTC: false, _pf: {…}, …}
originalStoryData: {id: 22, date: "2020-04-06T21:14:58", date_gmt: "2020-04-06T21:14:58", guid: {…}, modified: "2020-05-01T15:48:27", …}
pages: [{…}]
status: "draft"
tags: (3) [28, 29, 30]
title: "DERPIN WITH FRED"
}
```
Tags returned: 
Screen Shot 2020-05-01 at 9 13 06 AM

File: assets/src/dashboard/app/views/shared/storyListView.js [hunk 2]
```diff
@@ -92,13 +97,33 @@ const toggleSortLookup = {
...
+
 export default function StoryListView({
   filteredStories,
   storySort,
   handleSortChange,
   handleSortDirectionChange,
   sortDirection,
+  tags,
+  categories,
+  users,
```
Comment (BrittanyIRL) [inline assets/src/dashboard/app/views/shared/storyListView.js:121]: Curious your thoughts on just grabbing these from context here instead of passing them as props from the my-stories view since they are just objects getting referenced by Ids found in the `filteredStories` (which we should probably rename now that its not doing any kind of filtering, but that's beside the point).

**Review by BrittanyIRL — CHANGES_REQUESTED** [review]
I'm stoked you found a straight forward way to add in categories and tags! I think I found an edge case about mismatched ids. Let me know if you want more info - all i did to test this was go the the wordpress quick edit view for stories and add in tags and  categories hit 'update'.

File: assets/src/dashboard/app/api/useCategoriesApi.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ * Internal dependencies
+ */
+import groupBy from '../../utils/groupBy';
+
+export default function useCategoriesApi(dataAdapter, { wpApi }) {
+  const [categories, setCategories] = useState({});
+  const fetchCategories = useCallback(async () => {
+    try {
+      const response = await dataAdapter.get(wpApi, {
+        parse: false,
```
Comment (mariano-formidable) [inline assets/src/dashboard/app/api/useCategoriesApi.js]: I believe `parse: false` is only used by the story API so we could read the response header tags to get some extra info.  In this case, it looks like you're taking the response and immediately converting it to `json`, so can do it in one step like this:
```
const categoriesJson = await dataAdapter.get(wpApi);
```

File: assets/src/dashboard/app/api/useTagsApi.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import groupBy from '../../utils/groupBy';
+
+export default function useTagsApi(dataAdapter, { wpApi }) {
+  const [tags, setTags] = useState({});
+  const fetchTags = useCallback(async () => {
+    try {
+      const response = await dataAdapter.get(wpApi, {
+        parse: false,
+      });
+      const tagsJson = await response.json();
```
Comment (mariano-formidable) [inline assets/src/dashboard/app/api/useTagsApi.js]: Same here, this can be written:
```
const tagsJson = await dataAdapter.get(wpApi);
```

File: assets/src/dashboard/app/api/useUserApi.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import groupBy from '../../utils/groupBy';
+
+export default function useUsersApi(dataAdapter, { wpApi }) {
+  const [users, setUsers] = useState({});
+  const fetchUsers = useCallback(async () => {
+    try {
+      const response = await dataAdapter.get(wpApi, {
+        parse: false,
+      });
+      const usersJson = await response.json();
```
Comment (mariano-formidable) [inline assets/src/dashboard/app/api/useUserApi.js]: Also, this can be written:
```
const usersJson = await dataAdapter.get(wpApi);
```

File: assets/src/dashboard/app/views/shared/storyListView.js (hunk 1 above)
Comment (mariano-formidable) [reply assets/src/dashboard/app/views/shared/storyListView.js]: Yeah, perhaps this could be fixed with a `?` and `Boolean` filter like so:
```
.reduce((memo, current) => {
  return [...memo, metadata[current]?.name];
}, [])
.filter(Boolean)
.join(', ');
```

**Review by mariano-formidable — CHANGES_REQUESTED** [review]
Only a few nits, but overall looking really good!

**Review by mariano-formidable — APPROVED** [review]
Lookin' good!  Nice job 🎉

File: assets/src/dashboard/app/views/shared/storyListView.js [hunk 6]
```diff
@@ -92,13 +97,32 @@ const toggleSortLookup = {
...
 }) {
+  const metadataStringForIds = useCallback((metadata, ids) => {
+    const metadataString = ids
+      .reduce((memo, current) => [...memo, metadata[current]?.name], [])
```
Comment (BrittanyIRL) [inline assets/src/dashboard/app/views/shared/storyListView.js:125]: 🎉

File: assets/src/dashboard/app/views/shared/storyListView.js [hunk 7]
```diff
@@ -188,6 +213,9 @@ export default function StoryListView({
 
 StoryListView.propTypes = {
   filteredStories: StoriesPropType,
+  tags: PropTypes.object,
+  categories: PropTypes.object,
+  users: PropTypes.object,
```
Comment (BrittanyIRL) [inline assets/src/dashboard/app/views/shared/storyListView.js]: nit: add a shape so we know what's expected to be in here

**Review by BrittanyIRL — APPROVED** [review]
This works so well, great work around for the pagination issue with the tags and categories. And the table's standing up against really ridiculously long tags as well. One nit to add a shape for proptypes but nothing that'd stop merging.

File: assets/src/dashboard/app/api/fetchAllFromPages.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+ * @param wpApi The base url for the fetch calls
+ * @return Promise<any[]> A flattened array with all the entities
+ */
+
+export default async function fetchAllFromTotalPages(
+  response,
+  dataAdapter,
+  wpApi
+) {
+  const totalPages = parseInt(response.headers.get('X-WP-TotalPages'));
```
Comment (swissspidy) [inline assets/src/dashboard/app/api/fetchAllFromPages.js:38]: FWIW, the api-fetch package actually has a middleware that already does exactly this. There should be no need for it here.
